aQute.libg/src/aQute/lib/tag/Tag.java

```diff
@@ -131,6 +137,84 @@ public Tag addContent(Tag tag) {
 		return this;
 	}
 
+	/**
+	 * Add new content tags derived from the DTO value.
+	 */
+	public Tag addContent(String name, Object dto) {
+		if (dto != null) {
+			if (isComplex(dto)) {
+				List<Object> flattedDtos = new LinkedList<>();
+
+				if (flatCollection(dto, flattedDtos, true)) {
+					for (Object d : flattedDtos) {
+						addContent(name, d);
+					}
+				} else if (dto instanceof Map) {
+					Tag tag = new Tag(name);
+					for (Entry< ? , ? > entry : ((Map< ? , ? >) dto).entrySet()) {
+						tag.addContent(Objects.toString(entry.getKey()), entry.getValue());
+					}
+					if (!tag.content.isEmpty()) {
+						addContent(tag);
+					}
+				} else {
+					Tag tag = new Tag(name);
+					for (Field field : getFields(dto.getClass())) {
+						try {
+							tag.addContent(field.getName(), field.get(dto));
+						} catch (IllegalAccessException bug) {
+							/* should not be thrown if input respect dto spec */
+							throw new RuntimeException(bug);
+						}
+					}
+					addContent(tag);
+				}
+			} else {
+				addContent(new Tag(name, dto.toString()));
+			}
+		}
+		return this;
+	}
+
+	private boolean flatCollection(Object dto, List<Object> result, boolean topLevel) {
+		if (dto != null) {
+			if (dto.getClass().isArray()) {
+				int length = Array.getLength(dto);
+				for (int i = 0; i < length; i++) {
+					flatCollection(Array.get(dto, i), result, false);
+				}
+
+				return true;
+			} else if (dto instanceof Collection) {
+				for (Object d : (Collection< ? >) dto) {
+					flatCollection(d, result, false);
+				}
+
+				return true;
+			} else if (!topLevel) {
+				result.add(dto);
+			}
+		}
+		return false;
+	}
+
+	private boolean isComplex(Object a) {
+		return a instanceof Map || a instanceof Collection || a.getClass().isArray()
+				|| getFields(a.getClass()).length > 0;
+	}
+
+	private Field[] getFields(Class< ? > c) {
+		List<Field> publicFields = new ArrayList<>();
+
+		for (Field field : c.getFields()) {
+			if (field.isEnumConstant() || field.isSynthetic() || Modifier.isStatic(field.getModifiers()))
+				continue;
+			publicFields.add(field);
+		}
+
+		return publicFields.toArray(new Field[publicFields.size()]);
```

Comment:
Using size in the array allocation as input to `toArray` is bad practice for performance reasons. Use `0` as the size.